src/fd.cpp

```diff
@@ -0,0 +1,147 @@
+#ifdef _WIN32
+#include <winsock2.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include "later.h"
+#include "callback_registry_table.h"
+
+extern CallbackRegistryTable callbackRegistryTable;
+
+typedef struct ThreadArgs_s {
+  SEXP callback;
+  R_xlen_t num_fds;
+  std::unique_ptr<std::vector<int>> fds;
+  fd_set read_fds;
+  struct timeval tv;
+  int max_fd;
+  int loop;
+  bool flag;
+} ThreadArgs;
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  SEXP results, call;
+  // TODO: actually pre-allocate call on main thread, values can be safely updated from background thread
+  // only call Rf_eval() here to minimise runtime in later callback.
+  PROTECT(results = Rf_allocVector(LGLSXP, args->num_fds));
+  std::memcpy((void *) DATAPTR_RO(results), args->fds->data(), args->num_fds * sizeof(int));
+  PROTECT(call = Rf_lcons(args->callback, Rf_cons(results, R_NilValue)));
+  Rf_eval(call, R_GlobalEnv);
+  UNPROTECT(2);
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: upgrade select() to poll() / WSAPoll().
+// CONSIDER: add method for HANDLES on Windows. Assuming we only accept integer
+// values for both, we could use heuristics: check if it is a valid SOCKET or
+// else assume HANDLE - but this is not fool-proof.
+// Otherwise would require an interface for user to specify type.
+static void *select_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  // CONSIDER: if we should be checking more than read activity, i.e. write and error.
+  // Could check all types for all fds, which would be inefficient but
+  // otherwise would require user interface to specify events for each fd
+  int ready = select(args->max_fd + 1, &args->read_fds, NULL, NULL, args->flag ? &args->tv : NULL);
```

Comment:
I forgot that `select` only works for file descriptors <= 1024, another reason to at least use `poll()`.

Reply:
I've upgraded to `poll()` on Linux/Mac and the equivalent `WSAPoll()` on Windows. However the CI is now failing for R <= 4.1. Seems `WSAPoll()` might be missing from the 'Winsock2.h' header - I'll have to investigate further.

Reply:
I've re-introduced `select()` as a fallback for those older R versions on Windows, and increased the defined max file descriptors in that case to 2048. The checks are passing cleanly again! FWIW there's no discernible difference between `select()` and `poll()` for small numbers of FDs at least.